Release 2.0.0

lnedry opened the pull request:
Changes to config/bounce.ini
- removed check.non_local_msgid
- removed reject.non_local_msgid
- added check.hash_validation
- added check.hash_date
- added reject.hash_validation
- added reject.hash_date
- added validation.max_hash_age_days
- added validation.hash_algorithm
- added validation.secret

Added config/bounce_whitelist.json
- combinations of sender/recipient that should skip validation

Changes to test/index.js
- removed requirement for haraka-email-message
- removed tests for non_local_msgid
- added tests for create_validation_hash
- added tests for validate_bounce
- added tests for get_bounce_headers
- added tests for should_skip
- added tests for find_received_headers
- added tests for validate_hash_date
- replaced most sinon.stub with sinon.spy to better monitor code behavior
- refactored many tests to make better use of sinon spies and assertions

Changes to index.js
- removed load_host_list()
- removed load_allowed_msgid_domains()
- removed non_local_msgid()
- removed find_message_id_headers()
- added create_validation_hash - creates a cryptographic validation hash for outbound emails
- added validate_date - reject if original email was sent before the configured number of days
- added validate_bounce - validates the bounce by comparing the hash created on the original email
- added find_bounce_headers - grab Date, From, Message-ID, and hash headers from the body of the bounce
- added should_skip - determines whether validation checks should be skipped
- added is_whitelisted - checks to see if a From:To pair is whitelisted
- added checks to only register a hook if it is configured
- added descriptive messages to all transaction.results

Changes to package.json
- added devDependencies (mocha and prettier)
- removed dependency for haraka-tld
- removed dependency for haraka-email-message

Changes to README.md
- removed documentation for non_local_msgid
- added documentation for the validation header
- added documentation for the validation_date

Checklist:

- [X] docs updated
- [X] tests updated
- [ ] Changes.md updated
- [X] package.json.version bumped

lnedry committed:
Update bounce.ini

- added check.hash_validation
- added check.hash_date
- added reject.hash_validation
- added reject.hash_date
- added validation.max_hash_age_days
- added validation.hash_algorithm
- added validation.secret

- removed check.non_local_msgid
- removed reject.non_local_msgid

lnedry committed:
Update index.js

- added tests for create_validation_hash
- added tests for validate_bounce
- added tests for get_bounce_headers
- added tests for should_skip
- added tests for find_received_headers
- added tests for validate_hash_date

- removed requirement for haraka-email-message
- removed tests for non_local_msgid

- replaced most sinon.stub with sinon.spy to better monitor code behavior
- refactored many tests to make better use of sinon spies and assertions

lnedry committed:
Update README.md

- added documentation for the validation header
- added documentation for the validation_date

- removed documentation for non_local_msgid

lnedry committed:
Update index.js

- added checks to only register a hook if it is configured
- added create_validation_hash - creates a cryptographic validation hash for outbound emails
- added validate_date - reject if original email was sent before the configured number of days
- added validate_bounce - validates the bounce by comparing the hash created on the original email
- added find_bounce_headers - grab Date, From, Message-ID, and hash headers from the body of the bounce
- added should_skip - determines whether validation checks should be skipped
- added is_whitelisted - checks to see if a From:To pair is whitelisted
- added descriptive messages to all transaction.results

- removed load_host_list()
- removed load_allowed_msgid_domains()
- removed non_local_msgid()
- removed find_message_id_headers()

lnedry committed:
Update package.json

- added devDependencies (mocha and prettier)

- removed dependency for haraka-tld
- removed dependency for haraka-email-message

msimerson commented on index.js:
```diff
@@ -1,49 +1,28 @@
+
+  if (this.cfg.reject.all_bounces) this.register_hook('mail', 'reject_all')
+  if (this.cfg.reject.bad_rcpt) this.register_hook('rcpt_ok', 'bad_rcpt')
+  if (this.cfg.check.single_recipient)
```
The advantage of conditionally loading the hook here is that it saves Haraka from calling the hook on every connection. The disadvantage is that the function isn't enabled automatically when the config file is edited. You have to restart Haraka afterwards for the change to take effect. Since computing is so cheap, we usually opt to optimize for humans.

lnedry replied:
Fixed

msimerson commented on test/index.js:
```diff
@@ -1,349 +1,241 @@
-  }
+  this.connection.relaying = false
+  this.connection.transaction = fixtures.transaction.createTransaction()
+  this.connection.transaction.results = new fixtures.results(this.connection)
```
II think newer (less than 2 years old) versions of test-fixtures provide .results for us. Please confirm and remove this line.

lnedry replied:
I replaced that line with:
  this.connection.init_transaction()

msimerson reviewed (changes requested):
This looks great, but I'd like one more test added: call every hook with a bare connection (no transaction), and assure that the functions do something reasonable.

lnedry committed:
Update index.js

Removed conditional hook registrations

lnedry committed:
Update index.js

- fixed initialization of Results
- updated tests for register_hooks
- added tests for transaction guarding for each function

lnedry commented:
Changes to index.js
- Removed conditional hook registrations

Changes to test/index.js
- fixed initialization of Results
- updated tests for register_hooks
- added tests for transaction guarding for each function

msimerson reviewed (approved):
Looks Good To Me!
